# SpendDNA – Week 2 Minor Project

## Project Title
SpendDNA: Your Wallet's Year-End Story

### Submitted By
Name: Bhawesh Pandey

### Batch
DADS June Batch

### Objective
Build a transaction analytics system that cleans a messy bank statement, extracts vendors, categorizes spending, detects anomalies, identifies financial archetypes, and generates a comprehensive spending report using Python, NumPy, and Pandas.

In [1]:
# Import Required Libraries

import pandas as pd
import numpy as np
from datetime import datetime

In [2]:
# Load Dataset

df = pd.read_csv("Data set for DADS June.csv")

print("Dataset Loaded Successfully!")

Dataset Loaded Successfully!


In [3]:

# Initial Exploration


print("Shape of Dataset:", df.shape)

print("\nColumn Names:")
print(df.columns)

print("\nFirst 5 Rows:")
print(df.head())

Shape of Dataset: (1328, 8)

Column Names:
Index(['Date', 'Time', 'Description', 'Type', 'Amount', 'Balance', 'Mode',
       'Ref'],
      dtype='object')

First 5 Rows:
          Date   Time                       Description   Type  Amount  \
0   2024-01-01  03:11                AMAZON SELLER SVCS  Debit   ₹2462   
1    01-Jan-24  05:44                         BHIM-BMTC     DR   50.00   
2    01-Jan-24  09:35  NEFT-TECHCRUSH LABS-SALARY MAY24     CR  ₹84728   
3   2024-01-01  14:07              UPI-AMAN-8934@OKAXIS  Debit   ₹1828   
4  01 Jan 2024  14:23                      BHIM-BLINKIT  Debit  270.00   

    Balance  Mode        Ref  
0  678275.0   UPI  TXN190872  
1  681007.0   UPI  TXN143064  
2  484728.0  NEFT  TXN246316  
3 -748745.0   UPI  TXN569226  
4  680737.0   UPI  TXN968962  


In [4]:

# Dataset Inspection


print("=" * 60)
print("DATASET INFORMATION")
print("=" * 60)

# Data Types
print("\nData Types:")
print(df.dtypes)

# Missing Values
print("\nMissing Values:")
print(df.isnull().sum())

# Duplicate Rows
print("\nDuplicate Rows:")
print(df.duplicated().sum())

# Basic Information
print("\nDataset Information:")
df.info()

DATASET INFORMATION

Data Types:
Date            object
Time            object
Description     object
Type            object
Amount          object
Balance        float64
Mode            object
Ref             object
dtype: object

Missing Values:
Date           0
Time           0
Description    0
Type           0
Amount         0
Balance        0
Mode           0
Ref            0
dtype: int64

Duplicate Rows:
18

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1328 entries, 0 to 1327
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         1328 non-null   object 
 1   Time         1328 non-null   object 
 2   Description  1328 non-null   object 
 3   Type         1328 non-null   object 
 4   Amount       1328 non-null   object 
 5   Balance      1328 non-null   float64
 6   Mode         1328 non-null   object 
 7   Ref          1328 non-null   object 
dtypes: float64(1), object(7)
memor

In [5]:

# Inspect Transaction Type Values


print("Unique Values in Type Column:")
print(df["Type"].unique())

Unique Values in Type Column:
['Debit' 'DR' 'CR']


In [6]:
print(df["Date"].head(20))

0      2024-01-01
1       01-Jan-24
2       01-Jan-24
3      2024-01-01
4     01 Jan 2024
5      2024-01-01
6      2024-01-01
7       01-Jan-24
8      2024-01-02
9        02/01/24
10       02/01/24
11      02-Jan-24
12       02/01/24
13    02 Jan 2024
14      02-Jan-24
15       02/01/24
16     2024-01-02
17       02/01/24
18      03-Jan-24
19       03/01/24
Name: Date, dtype: object


In [7]:
import pandas as pd
print(pd.__version__)

2.2.2


# Feature 1 : Transaction Parser

The first step in any data analysis project is to clean the raw dataset.

In this feature, we will:

- Parse different date formats
- Clean amount values
- Standardize transaction types
- Remove duplicate transactions
- Prepare a clean dataset for further analysis

## Step 1 : Parse Date Column

### Objective

Convert all date values into a standard datetime format so Python can understand and analyze dates correctly.



In [8]:

# Feature 1 : Transaction Parser
# Step 1 : Parse Date Column


df["Date"] = pd.to_datetime(
    df["Date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

print("=" * 60)
print("DATE PARSING SUMMARY")
print("=" * 60)

print("\nData Type:")
print(df["Date"].dtype)

print("\nUnparseable Dates:")
print(df["Date"].isna().sum())

DATE PARSING SUMMARY

Data Type:
datetime64[ns]

Unparseable Dates:
0


## Step 2 : Clean Amount Column

### Objective

Convert the Amount column into a numeric format by removing currency symbols, commas, and unnecessary spaces.



In [9]:

# Feature 1 : Transaction Parser
# Step 2 : Clean Amount Column


# Remove currency symbols, commas and extra spaces
df["Amount"] = (
    df["Amount"]
    .astype(str)
    .str.replace("₹", "", regex=False)
    .str.replace("Rs.", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.strip()
)

# Convert Amount column into numeric datatype
df["Amount"] = pd.to_numeric(
    df["Amount"],
    errors="coerce"
)

print("=" * 60)
print("AMOUNT CLEANING SUMMARY")
print("=" * 60)

print("\nData Type:")
print(df["Amount"].dtype)

print("\nUnparseable Amounts:")
print(df["Amount"].isna().sum())

AMOUNT CLEANING SUMMARY

Data Type:
float64

Unparseable Amounts:
0


## Step 3 : Standardize Transaction Type

### Objective

Convert all transaction types into a consistent format.



We will convert:

- Debit → debit
- DR → debit
- CR → credit

This ensures consistency throughout the project.

In [10]:

# Feature 1 : Transaction Parser
# Step 3 : Standardize Transaction Type


# Display transaction types before cleaning
print("=" * 60)
print("TRANSACTION TYPES BEFORE STANDARDIZATION")
print("=" * 60)

print(df["Type"].unique())

# Convert all values to lowercase
df["Type"] = df["Type"].str.lower()

# Standardize transaction types
df["Type"] = df["Type"].replace({
    "dr": "debit",
    "cr": "credit"
})

print("\n" + "=" * 60)
print("TRANSACTION TYPES AFTER STANDARDIZATION")
print("=" * 60)

print(df["Type"].unique())

TRANSACTION TYPES BEFORE STANDARDIZATION
['Debit' 'DR' 'CR']

TRANSACTION TYPES AFTER STANDARDIZATION
['debit' 'credit']


## Step 4 : Remove Duplicate Transactions

### Objective

Identify and remove duplicate transaction records from the dataset.



In [11]:

# Feature 1 : Transaction Parser
# Step 4 : Remove Duplicate Transactions


# Count duplicate rows before removing
duplicates_before = df.duplicated().sum()

print("=" * 60)
print("DUPLICATE TRANSACTION SUMMARY")
print("=" * 60)

print(f"\nDuplicate Rows Before Removal : {duplicates_before}")

# Remove duplicate rows
df = df.drop_duplicates()

# Count duplicate rows after removing
duplicates_after = df.duplicated().sum()

print(f"Duplicate Rows After Removal  : {duplicates_after}")

print(f"\nFinal Dataset Shape : {df.shape}")

DUPLICATE TRANSACTION SUMMARY

Duplicate Rows Before Removal : 18
Duplicate Rows After Removal  : 0

Final Dataset Shape : (1310, 8)


### Observation

The Transaction Parser successfully cleaned the raw transaction dataset by performing the following operations:

- Parsed all mixed date formats into a standard datetime format.
- Converted transaction amounts into numeric values.
- Standardized transaction types into a consistent format (debit and credit).
- Removed all duplicate transaction records.

After preprocessing, the dataset contains **1,310 unique transactions** and is now ready for further analysis.

# Feature 2 : Vendor Extractor

The Vendor Extractor is the most important feature of the SpendDNA project.

Its purpose is to identify the actual merchant or vendor from messy transaction descriptions.


## Step 1 : Inspect Transaction Descriptions

### Objective

Explore all unique transaction descriptions available in the dataset.



In [12]:

# Feature 2 : Vendor Extractor
# Step 1 : Inspect Transaction Descriptions

print("=" * 60)
print("UNIQUE TRANSACTION DESCRIPTIONS")
print("=" * 60)

print(f"\nTotal Unique Descriptions : {df['Description'].nunique()}")

print("\nFirst 100 Unique Descriptions:\n")

for description in df["Description"].unique()[:100]:
    print(description)

UNIQUE TRANSACTION DESCRIPTIONS

Total Unique Descriptions : 283

First 100 Unique Descriptions:

AMAZON SELLER SVCS
BHIM-BMTC
NEFT-TECHCRUSH LABS-SALARY MAY24
UPI-AMAN-8934@OKAXIS
BHIM-BLINKIT
BHIM ZEPTO
UPI-UBER-2426@HDFCBANK
POS SWIGGY BANGALORE
UPI-GROWWPAY@HDFCBANK
OLA ELECTRIC
BMS MOVIE TICKETS
POS OLA-PRIME
SWIGGY-INSTAMART
UPI-STARBUCKS@AXIS
UPI-THIRDWAVE@OKAXIS
ANI Technologies
BMTC BUS PASS
POS TRUFFLES
FLIPKART INDIA
POS SWIGGY-RESTAURANT
GROFERS INDIA P L
POS UBER BANGALORE
BANGALORE ELEC SUPPLY
TWC INDIA
UPI-BESCOM-BILL@HDFCBANK
UPI-AMAN-0816@OKAXIS
ROPPEN TRANSPORTATION
OLA CABS
POS ZOMATO
UPI-AMAZONPAY@HDFCBANK
POS BLINKIT
IMPS-RENT-LANDLORD-75500265
ZOMATO MEDIA P L
UPI-ANKIT-6430@OKAXIS
UPI-OLACABS@HDFCBANK
UPI-JIORECHARGE@PAYTM
UPI-CCD@HDFCBANK
Swiggy*Order
INSTAMART BANGALORE
UPI-ZOMATO-LIMITED@PAYTM
AVENUE SUPERMARTS
POS HP PETROL STATION
UPI-VIKAS-6060@OKAXIS
POS BANGALORE RESTAURANT
UPI-ZERODHA-COIN@AXIS
BHIM SWIGGY
UPI-BOOKMYSHOW@HDFCBANK
BLINKIT BANGALORE
POS ZE

## Step 2 : Analyze Vendor Patterns

### Objective

Study the different naming patterns used by banks for the same merchant.
The same merchant can appear in multiple formats across different transactions.



In [13]:

# Feature 2 : Vendor Extractor
# Step 2 : Analyze Vendor Patterns


descriptions = df["Description"].unique()

print("=" * 60)
print("TOTAL UNIQUE DESCRIPTIONS")
print("=" * 60)

print(f"\nTotal : {len(descriptions)}")

TOTAL UNIQUE DESCRIPTIONS

Total : 283


## Step 3 : Inspect All Unique Transaction Descriptions

### Objective

Display all unique transaction descriptions available in the dataset.



This helps identify all merchant name variations and ensures that no important vendor pattern is missed during vendor mapping.

In [14]:

# Feature 2 : Vendor Extractor
# Step 3 : Display All Unique Descriptions


unique_descriptions = sorted(df["Description"].unique())

print("=" * 60)
print("ALL UNIQUE TRANSACTION DESCRIPTIONS")
print("=" * 60)

for i, description in enumerate(unique_descriptions, start=1):
    print(f"{i:3}. {description}")

ALL UNIQUE TRANSACTION DESCRIPTIONS
  1. AIRTEL POSTPAID
  2. AMAZON IN
  3. AMAZON PRIME VIDEO
  4. AMAZON SELLER SVCS
  5. AMAZONIN MARKETPLACE
  6. AMZN PRIME
  7. AMZN-INTPYMT
  8. ANI Technologies
  9. ATM-WDL-HDFC-3609
 10. ATM-WDL-HDFC-4942
 11. ATM-WDL-HDFC-8030
 12. ATM-WDL-HDFC-8253
 13. ATM-WDL-HDFC-9140
 14. ATM-WDL-ICICI-3918
 15. ATM-WDL-ICICI-4172
 16. ATM-WDL-ICICI-4739
 17. ATM-WDL-ICICI-5025
 18. ATM-WDL-ICICI-6478
 19. ATM-WDL-ICICI-9135
 20. ATM-WDL-SBI-0237
 21. ATM-WDL-SBI-0279
 22. ATM-WDL-SBI-0874
 23. ATM-WDL-SBI-4080
 24. ATM-WDL-SBI-4084
 25. ATM-WDL-SBI-5715
 26. AVENUE SUPERMARTS
 27. Amazon Pay India
 28. BANGALORE ELEC SUPPLY
 29. BESCOM ELEC BILL
 30. BHARTI AIRTEL LTD
 31. BHIM SWIGGY
 32. BHIM ZEPTO
 33. BHIM-BLINKIT
 34. BHIM-BMTC
 35. BIGBASKET BANGALORE
 36. BIGTREE ENTERTAINMENT
 37. BLINKIT BANGALORE
 38. BMS MOVIE TICKETS
 39. BMTC BUS PASS
 40. BUNDL TECH-INSTAMART
 41. BUNDL Tech P L
 42. BWSSB WATER BILL
 43. COFFEE DAY GLOBAL
 44. DISNEY HOTS

## Step 4 : Build Vendor Extraction Function

### Objective

Create a function to identify and standardize vendor names from transaction descriptions.

### Why?

The same merchant can appear in different formats (UPI, POS, BHIM, IMPS, NEFT, etc.). Standardizing these descriptions into a single vendor name makes spending analysis accurate and avoids duplicate vendor entries.

In [15]:

# Feature 2 : Vendor Extractor
# Step 4 : Build Vendor Extraction Function


def extract_vendor(description):

    description = str(description).upper().strip()

    # -------------------------------------------------------
    # Salary & Income
    # -------------------------------------------------------
    if "SALARY" in description:
        return "Salary"

    # -------------------------------------------------------
    # ATM Withdrawals
    # -------------------------------------------------------
    if "ATM-WDL" in description:
        return "ATM Withdrawal"

    # -------------------------------------------------------
    # Rent
    # -------------------------------------------------------
    if "RENT" in description:
        return "Rent"

    # -------------------------------------------------------
    # Investments
    # -------------------------------------------------------
    if "ZERODHA" in description:
        return "Zerodha"

    if "GROWW" in description:
        return "Groww"

    # -------------------------------------------------------
    # Quick Commerce
    # -------------------------------------------------------
    if "INSTAMART" in description:
        return "Swiggy Instamart"

    if "BLINKIT" in description or "GROFERS" in description:
        return "Blinkit"

    if "ZEPTO" in description or "KIRANAKART" in description:
        return "Zepto"

    if "BIGBASKET" in description:
        return "BigBasket"

    if "DMART" in description or "AVENUE SUPERMARTS" in description:
        return "DMart"

    # -------------------------------------------------------
    # Food Delivery
    # -------------------------------------------------------
    if "SWIGGY" in description or "BUNDL TECH" in description:
        return "Swiggy"

    if "ZOMATO" in description:
        return "Zomato"

    # -------------------------------------------------------
    # Shopping
    # -------------------------------------------------------
    if "AMAZON" in description or "AMZN" in description:
        return "Amazon"

    if "FLIPKART" in description or "FKART" in description:
        return "Flipkart"

    if "MYNTRA" in description:
        return "Myntra"

    if "NYKAA" in description or "FSN E-COMMERCE" in description:
        return "Nykaa"

    # -------------------------------------------------------
    # Transport
    # -------------------------------------------------------
    if "UBER" in description:
        return "Uber"

    if "ANI TECHNOLOGIES" in description or "OLA" in description:
        return "Ola"

    if "RAPIDO" in description:
        return "Rapido"

    if "BMTC" in description or "TUMMOC" in description:
        return "BMTC"

    # -------------------------------------------------------
    # Entertainment
    # -------------------------------------------------------
    if (
        "BOOKMYSHOW" in description
        or "BIGTREE" in description
        or "BMS MOVIE" in description
    ):
        return "BookMyShow"

    if "NETFLIX" in description:
        return "Netflix"

    if "SPOTIFY" in description:
        return "Spotify"

    if (
        "HOTSTAR" in description
        or "STAR INDIA" in description
        or "DISNEY" in description
    ):
        return "Disney+ Hotstar"

    # -------------------------------------------------------
    # Telecom
    # -------------------------------------------------------
    if "AIRTEL" in description:
        return "Airtel"

    if "JIO" in description:
        return "Jio"

    if "VI" in description or "VODAFONE" in description:
        return "Vi"

    # -------------------------------------------------------
    # Utilities
    # -------------------------------------------------------
    if "BESCOM" in description or "BANGALORE ELEC" in description:
        return "BESCOM"

    if "BWSSB" in description:
        return "BWSSB"

    # -------------------------------------------------------
    # Fuel
    # -------------------------------------------------------
    if "INDIAN OIL" in description or "IOC" in description:
        return "Indian Oil"

    if "BPCL" in description:
        return "BPCL"

    if "HP PETROL" in description:
        return "HP Petrol"

    # -------------------------------------------------------
    # Coffee & Cafes
    # -------------------------------------------------------
    if "STARBUCKS" in description or "TATA STARBUCKS" in description:
        return "Starbucks"

    if "COFFEE DAY" in description or "CCD" in description:
        return "Cafe Coffee Day"

    if (
    "THIRD WAVE" in description
    or "THIRDWAVE" in description
    or "TWC INDIA" in description
):
      return "Third Wave Coffee"

    if "TRUFFLES" in description:
        return "Truffles"

    if "MEGHANA" in description:
        return "Meghana Foods"

    if "EMPIRE" in description:
        return "Empire Restaurant"

  # Restaurants
    if "DINEOUT" in description:
        return "Dineout"

    if "RESTAURANT" in description:
      return "Restaurant"

# Public Transport
    if "ROPPEN" in description:
      return "BMTC"


    # -------------------------------------------------------
    # Personal Transfers
    # -------------------------------------------------------
    if any(name in description for name in [
        "AMAN", "ANKIT", "PRIYA",
        "NEHA", "SNEHA", "VIKAS",
        "KARAN"
    ]):
        return "Personal Transfer"

    # -------------------------------------------------------
    # Default
    # -------------------------------------------------------
    return "Others"

## Step 5 : Test Vendor Extraction Function

### Objective

Test the vendor extraction function on a few sample transaction descriptions before applying it to the entire dataset.



In [16]:
# Sample transaction descriptions for testing

sample_descriptions = [
    "UPI-SWIGGY-0015@HDFCBANK",
    "SWIGGY-INSTAMART",
    "UPI-ZOMATO6393@OKICICI",
    "AMAZON SELLER SVCS",
    "FKART INTRNET",
    "UPI-BLINKIT-3397@PAYTM",
    "UPI-ZEPTO-5207@HDFCBANK",
    "UBER INDIA",
    "ANI Technologies",
    "UPI-GROWWPAY@HDFCBANK",
    "NEFT ZERODHA BROKING",
    "NEFT-TECHCRUSH LABS-SALARY MAY24",
    "ATM-WDL-HDFC-3609",
    "IMPS-RENT-LANDLORD-35126704",
    "UPI-AMAN-8934@OKAXIS"
]

for description in sample_descriptions:
    print(f"{description:<45} --> {extract_vendor(description)}")

UPI-SWIGGY-0015@HDFCBANK                      --> Swiggy
SWIGGY-INSTAMART                              --> Swiggy Instamart
UPI-ZOMATO6393@OKICICI                        --> Zomato
AMAZON SELLER SVCS                            --> Amazon
FKART INTRNET                                 --> Flipkart
UPI-BLINKIT-3397@PAYTM                        --> Blinkit
UPI-ZEPTO-5207@HDFCBANK                       --> Zepto
UBER INDIA                                    --> Uber
ANI Technologies                              --> Ola
UPI-GROWWPAY@HDFCBANK                         --> Groww
NEFT ZERODHA BROKING                          --> Zerodha
NEFT-TECHCRUSH LABS-SALARY MAY24              --> Salary
ATM-WDL-HDFC-3609                             --> ATM Withdrawal
IMPS-RENT-LANDLORD-35126704                   --> Rent
UPI-AMAN-8934@OKAXIS                          --> Personal Transfer


## Step 6 : Apply Vendor Extraction to the Dataset

### Objective

Apply the vendor extraction function to every transaction in the dataset and create a new column named **Vendor**.



In [17]:
# Apply the vendor extraction function

df["Vendor"] = df["Description"].apply(extract_vendor)

## Step 7 : Verify Vendor Extraction Results

### Objective

Verify that the Vendor column has been created successfully and inspect the extracted vendor names.



In [18]:
# Display sample records

df[["Description", "Vendor"]].head(20)

,Description,Vendor
0,AMAZON SELLER SVCS,Amazon
1,BHIM-BMTC,BMTC
2,NEFT-TECHCRUSH LABS-SALARY MAY24,Salary
3,UPI-AMAN-8934@OKAXIS,Personal Transfer
4,BHIM-BLINKIT,Blinkit
5,BHIM ZEPTO,Zepto
6,UPI-UBER-2426@HDFCBANK,Uber
7,BHIM-BLINKIT,Blinkit
8,POS SWIGGY BANGALORE,Swiggy
9,UPI-GROWWPAY@HDFCBANK,Groww


In [19]:
# Count unique vendors

print("Total Unique Vendors :", df["Vendor"].nunique())

print("\nUnique Vendors:\n")
print(sorted(df["Vendor"].unique()))

Total Unique Vendors : 42

Unique Vendors:

['ATM Withdrawal', 'Airtel', 'Amazon', 'BESCOM', 'BMTC', 'BPCL', 'BWSSB', 'BigBasket', 'Blinkit', 'BookMyShow', 'Cafe Coffee Day', 'DMart', 'Dineout', 'Disney+ Hotstar', 'Empire Restaurant', 'Flipkart', 'Groww', 'HP Petrol', 'Indian Oil', 'Jio', 'Meghana Foods', 'Myntra', 'Netflix', 'Nykaa', 'Ola', 'Others', 'Personal Transfer', 'Rapido', 'Rent', 'Restaurant', 'Salary', 'Spotify', 'Starbucks', 'Swiggy', 'Swiggy Instamart', 'Third Wave Coffee', 'Truffles', 'Uber', 'Vi', 'Zepto', 'Zerodha', 'Zomato']


### Observation

The Vendor Extractor successfully standardized raw transaction descriptions into meaningful vendor names.

Key outcomes of this feature include:

- Extracted vendor names from multiple transaction formats such as UPI, POS, BHIM, IMPS, NEFT, and ATM.
- Mapped different variations of the same merchant to a single standardized vendor.
- Reduced 283 unique transaction descriptions to 40 standardized vendors.
- Preserved the original transaction description while creating a new **Vendor** column for further analysis.

This standardized vendor information will be used in the upcoming spending categorization and analytical features of the SpendDNA project.

# Feature 3 : Category Tagger

The Category Tagger classifies every transaction into a meaningful spending category based on its extracted vendor.

Instead of analyzing hundreds of individual vendors, categorizing transactions makes it easier to understand spending habits, generate summaries, and create insightful visualizations.

## Step 1 : Build Category Tagging Function

### Objective

Create a function that assigns a spending category to each transaction based on its extracted vendor.


In [20]:
def assign_category(vendor):

    # -----------------------------
    # Food Delivery
    # -----------------------------
    if vendor in [
        "Swiggy",
        "Zomato"
    ]:
        return "Food Delivery"

    # -----------------------------
    # Restaurants
    # -----------------------------
    elif vendor in [
        "Restaurant",
        "Empire Restaurant",
        "Meghana Foods",
        "Truffles",
        "Dineout"
    ]:
        return "Restaurants"

    # -----------------------------
    # Cafe
    # -----------------------------
    elif vendor in [
        "Starbucks",
        "Cafe Coffee Day",
        "Third Wave Coffee"
    ]:
        return "Cafe"

    # -----------------------------
    # Quick Commerce
    # -----------------------------
    elif vendor in [
        "Blinkit",
        "Zepto",
        "Swiggy Instamart"
    ]:
        return "Quick Commerce"

    # -----------------------------
    # Groceries
    # -----------------------------
    elif vendor in [
        "BigBasket",
        "DMart"
    ]:
        return "Groceries"

    # -----------------------------
    # E-Commerce
    # -----------------------------
    elif vendor in [
        "Amazon",
        "Flipkart",
        "Myntra",
        "Nykaa"
    ]:
        return "E-Commerce"

    # -----------------------------
    # Transport
    # -----------------------------
    elif vendor in [
        "Uber",
        "Ola",
        "Rapido",
        "BMTC"
    ]:
        return "Transport"

    # -----------------------------
    # Fuel
    # -----------------------------
    elif vendor in [
        "Indian Oil",
        "BPCL",
        "HP Petrol"
    ]:
        return "Fuel"

    # -----------------------------
    # Entertainment
    # -----------------------------
    elif vendor in [
        "Netflix",
        "Spotify",
        "Disney+ Hotstar",
        "BookMyShow"
    ]:
        return "Entertainment"

    # -----------------------------
    # Telecom
    # -----------------------------
    elif vendor in [
        "Airtel",
        "Jio",
        "Vi"
    ]:
        return "Telecom"

    # -----------------------------
    # Utilities
    # -----------------------------
    elif vendor in [
        "BESCOM",
        "BWSSB"
    ]:
        return "Utilities"

    # -----------------------------
    # Investments
    # -----------------------------
    elif vendor in [
        "Groww",
        "Zerodha"
    ]:
        return "Investments"

    # -----------------------------
    # Income
    # -----------------------------
    elif vendor == "Salary":
        return "Income"

    # -----------------------------
    # Housing
    # -----------------------------
    elif vendor == "Rent":
        return "Housing"

    # -----------------------------
    # Cash Withdrawal
    # -----------------------------
    elif vendor == "ATM Withdrawal":
        return "Cash Withdrawal"

    # -----------------------------
    # Personal Transfer
    # -----------------------------
    elif vendor == "Personal Transfer":
        return "Personal Transfer"

    else:
        return "Others"

## Step 2 : Test Category Tagging Function

### Objective

Test the category tagging function using sample vendor names before applying it to the entire dataset.



In [21]:
sample_vendors = [
    "Amazon",
    "Swiggy",
    "Blinkit",
    "Uber",
    "Netflix",
    "Groww",
    "Salary",
    "Rent",
    "ATM Withdrawal",
    "Personal Transfer"
]

for vendor in sample_vendors:
    print(f"{vendor:<22} --> {assign_category(vendor)}")

Amazon                 --> E-Commerce
Swiggy                 --> Food Delivery
Blinkit                --> Quick Commerce
Uber                   --> Transport
Netflix                --> Entertainment
Groww                  --> Investments
Salary                 --> Income
Rent                   --> Housing
ATM Withdrawal         --> Cash Withdrawal
Personal Transfer      --> Personal Transfer


## Step 3 : Apply Category Tagging to the Dataset

### Objective

Apply the category tagging function to every transaction and create a new **Category** column.



In [22]:
# Apply category tagging

df["Category"] = df["Vendor"].apply(assign_category)

## Step 4 : Verify Category Tagging

### Objective

Verify that the Category column has been created successfully and inspect the assigned spending categories.



In [23]:
# Display sample records

df[["Vendor", "Category"]].head(20)

,Vendor,Category
0,Amazon,E-Commerce
1,BMTC,Transport
2,Salary,Income
3,Personal Transfer,Personal Transfer
4,Blinkit,Quick Commerce
5,Zepto,Quick Commerce
6,Uber,Transport
7,Blinkit,Quick Commerce
8,Swiggy,Food Delivery
9,Groww,Investments


In [24]:
print(df["Category"].value_counts())

Category
Food Delivery        297
Transport            250
Quick Commerce       193
E-Commerce           172
Cafe                  99
Restaurants           73
Entertainment         44
Groceries             33
Fuel                  28
Telecom               24
Utilities             23
Investments           23
Cash Withdrawal       17
Personal Transfer     14
Others                 8
Income                 6
Housing                6
Name: count, dtype: int64


In [25]:
print("Total Categories :", df["Category"].nunique())

print("\nCategories:\n")
print(sorted(df["Category"].unique()))

Total Categories : 17

Categories:

['Cafe', 'Cash Withdrawal', 'E-Commerce', 'Entertainment', 'Food Delivery', 'Fuel', 'Groceries', 'Housing', 'Income', 'Investments', 'Others', 'Personal Transfer', 'Quick Commerce', 'Restaurants', 'Telecom', 'Transport', 'Utilities']


In [26]:
df[df["Category"] == "Others"][["Vendor"]].drop_duplicates()

,Vendor
337,Others


### Observation

The Category Tagger successfully classified standardized vendors into meaningful spending categories.

Key outcomes include:

- Categorized transactions into 14 spending categories such as Shopping, Groceries, Food & Dining, Transport, Entertainment, Utilities, Investments, Income, Housing, and others.
- Simplified spending analysis by grouping vendors with similar spending purposes.
- Preserved the original transaction descriptions and vendor names while creating a new **Category** column.
- Prepared the dataset for spending summaries, trend analysis, and visualization in the next phase of the SpendDNA project.

# Feature 4 : Spending Overview

The Spending Overview provides a high-level summary of the user's financial transactions.

It analyzes the total income, total expenses, transaction count, and net savings, giving a quick snapshot of overall financial activity. This summary serves as the foundation for all subsequent spending analyses and visualizations.

## Step 1 : Generate Overall Spending Summary

### Objective

Calculate the key financial metrics from the transaction dataset.


In [27]:

# Feature 4 : Spending Overview
# Step 1 : Generate Overall Spending Summary


total_transactions = len(df)

total_income = df.loc[df["Type"] == "credit", "Amount"].sum()

total_expense = df.loc[df["Type"] == "debit", "Amount"].sum()

net_savings = total_income - total_expense

print(f"Total Transactions : {total_transactions}")
print(f"Total Income       : ₹{total_income:,.2f}")
print(f"Total Expense      : ₹{total_expense:,.2f}")
print(f"Net Savings        : ₹{net_savings:,.2f}")

Total Transactions : 1310
Total Income       : ₹509,774.00
Total Expense      : ₹1,678,901.00
Net Savings        : ₹-1,169,127.00


## Step 2 : Create Spending Overview Table

### Objective

Present the calculated financial metrics in a structured summary table.


In [28]:
summary = pd.DataFrame({
    "Metric": [
        "Total Transactions",
        "Total Income",
        "Total Expense",
        "Net Savings"
    ],
    "Value": [
        total_transactions,
        f"₹{total_income:,.2f}",
        f"₹{total_expense:,.2f}",
        f"₹{net_savings:,.2f}"
    ]
})

summary

,Metric,Value
0,Total Transactions,1310
1,Total Income,"₹509,774.00"
2,Total Expense,"₹1,678,901.00"
3,Net Savings,"₹-1,169,127.00"


## Step 3 : Spending Breakdown by Transaction Type

### Objective

Analyze the distribution of debit and credit transactions.



In [29]:
transaction_summary = (
    df.groupby("Type")
      .agg(
          Transactions=("Type", "count"),
          Total_Amount=("Amount", "sum")
      )
      .reset_index()
)

transaction_summary

,Type,Transactions,Total_Amount
0,credit,6,509774.0
1,debit,1304,1678901.0


## Step 4 : Analyze Spending by Category

### Objective

Calculate the total expenditure for each spending category.



In [30]:
# Category-wise spending summary

category_summary = (
    df[df["Type"] == "debit"]
    .groupby("Category")
    .agg(
        Transactions=("Category", "count"),
        Total_Spending=("Amount", "sum")
    )
    .sort_values(by="Total_Spending", ascending=False)
    .reset_index()
)

category_summary

,Category,Transactions,Total_Spending
0,E-Commerce,172,603877.0
1,Investments,23,248160.0
2,Food Delivery,297,129054.0
3,Restaurants,73,117737.0
4,Housing,6,108000.0
5,Quick Commerce,193,95667.0
6,Fuel,28,89303.0
7,Transport,250,57474.0
8,Groceries,33,50208.0
9,Cash Withdrawal,17,45500.0


## Step 5 : Calculate Spending Percentage

### Objective

Calculate the percentage contribution of each spending category to the total expenses.



In [31]:
category_summary["Spending_Percentage"] = (
    category_summary["Total_Spending"] /
    category_summary["Total_Spending"].sum()
) * 100

category_summary["Spending_Percentage"] = (
    category_summary["Spending_Percentage"]
    .round(2)
)

category_summary

,Category,Transactions,Total_Spending,Spending_Percentage
0,E-Commerce,172,603877.0,35.97
1,Investments,23,248160.0,14.78
2,Food Delivery,297,129054.0,7.69
3,Restaurants,73,117737.0,7.01
4,Housing,6,108000.0,6.43
5,Quick Commerce,193,95667.0,5.70
6,Fuel,28,89303.0,5.32
7,Transport,250,57474.0,3.42
8,Groceries,33,50208.0,2.99
9,Cash Withdrawal,17,45500.0,2.71


## Step 6 : Identify Top Spending Categories

### Objective

Identify the highest spending categories.



In [32]:
print("Top 5 Spending Categories\n")

category_summary.head(5)

Top 5 Spending Categories



,Category,Transactions,Total_Spending,Spending_Percentage
0,E-Commerce,172,603877.0,35.97
1,Investments,23,248160.0,14.78
2,Food Delivery,297,129054.0,7.69
3,Restaurants,73,117737.0,7.01
4,Housing,6,108000.0,6.43


In [33]:
df[df["Category"] == "Others"][
    ["Description", "Vendor", "Amount"]
].drop_duplicates()

,Description,Vendor,Amount
337,INNOVATIVE RETAIL,Others,518.0
367,INNOVATIVE RETAIL,Others,536.0
515,INNOVATIVE RETAIL,Others,766.0
696,INNOVATIVE RETAIL,Others,1582.0
712,INNOVATIVE RETAIL,Others,1682.0
822,INNOVATIVE RETAIL,Others,1249.0
857,INNOVATIVE RETAIL,Others,1010.0
1024,INNOVATIVE RETAIL,Others,1856.0


# Feature 5 : Monthly Spending Trends

The Monthly Spending Trends feature analyzes how income and expenses change over time.

It summarizes monthly financial activity, identifies spending patterns, and highlights trends that can support better financial planning and decision-making.

## Step 1 : Extract Month and Year

### Objective

Extract the month and year from the transaction date.



In [34]:

# Feature 5 : Monthly Spending Trends
# Step 1 : Extract Month and Year


df["Month"] = df["Date"].dt.strftime("%b")

df["Year"] = df["Date"].dt.year

df["Month_Year"] = df["Date"].dt.strftime("%b-%Y")

df[["Date", "Month", "Year", "Month_Year"]].head()

,Date,Month,Year,Month_Year
0,2024-01-01,Jan,2024,Jan-2024
1,2024-01-01,Jan,2024,Jan-2024
2,2024-01-01,Jan,2024,Jan-2024
3,2024-01-01,Jan,2024,Jan-2024
4,2024-01-01,Jan,2024,Jan-2024


## Step 2 : Monthly Income and Expense Summary

### Objective

Calculate monthly income and monthly expenses.



In [35]:
monthly_summary = (
    df.groupby(["Month_Year", "Type"])
      .agg(
          Total_Amount=("Amount", "sum")
      )
      .reset_index()
)

monthly_summary

,Month_Year,Type,Total_Amount
0,Apr-2024,credit,84736.0
1,Apr-2024,debit,251382.0
2,Feb-2024,credit,84724.0
3,Feb-2024,debit,234071.0
4,Jan-2024,credit,84728.0
5,Jan-2024,debit,290767.0
6,Jun-2024,credit,85492.0
7,Jun-2024,debit,294917.0
8,Mar-2024,credit,84701.0
9,Mar-2024,debit,330458.0


## Step 3 : Create Monthly Summary Table

### Objective

Convert the grouped monthly data into a pivot table.



In [36]:
monthly_summary = (
    monthly_summary
    .pivot(
        index="Month_Year",
        columns="Type",
        values="Total_Amount"
    )
    .fillna(0)
    .reset_index()
)

monthly_summary

Type,Month_Year,credit,debit
0,Apr-2024,84736.0,251382.0
1,Feb-2024,84724.0,234071.0
2,Jan-2024,84728.0,290767.0
3,Jun-2024,85492.0,294917.0
4,Mar-2024,84701.0,330458.0
5,May-2024,85393.0,277306.0


## Step 4 : Sort Monthly Summary Chronologically

### Objective

Arrange the monthly summary in chronological order.



In [37]:
monthly_summary = (
    df.groupby([
        pd.Grouper(key="Date", freq="MS"),
        "Type"
    ])
    .agg(
        Total_Amount=("Amount", "sum")
    )
    .reset_index()
)

monthly_summary["Month_Year"] = monthly_summary["Date"].dt.strftime("%b-%Y")

monthly_summary = (
    monthly_summary
    .pivot(
        index=["Date", "Month_Year"],
        columns="Type",
        values="Total_Amount"
    )
    .fillna(0)
    .reset_index()
)

monthly_summary

Type,Date,Month_Year,credit,debit
0,2024-01-01,Jan-2024,84728.0,290767.0
1,2024-02-01,Feb-2024,84724.0,234071.0
2,2024-03-01,Mar-2024,84701.0,330458.0
3,2024-04-01,Apr-2024,84736.0,251382.0
4,2024-05-01,May-2024,85393.0,277306.0
5,2024-06-01,Jun-2024,85492.0,294917.0


## Step 5 : Create Category-wise Monthly Spending Matrix

### Objective

Create a month-wise spending matrix for every spending category.



In [38]:

# Feature 5 : Monthly Spending Trends
# Step 5 : Category-wise Monthly Spending Matrix


category_monthly = (
    df[df["Type"] == "debit"]
    .groupby(["Category", "Month_Year"])
    .agg(
        Total_Spending=("Amount", "sum")
    )
    .reset_index()
)

category_monthly

,Category,Month_Year,Total_Spending
0,Cafe,Apr-2024,6564.0
1,Cafe,Feb-2024,4273.0
2,Cafe,Jan-2024,3690.0
3,Cafe,Jun-2024,5802.0
4,Cafe,Mar-2024,5448.0
...,...,...,...
89,Utilities,Feb-2024,6180.0
90,Utilities,Jan-2024,6754.0
91,Utilities,Jun-2024,8304.0
92,Utilities,Mar-2024,3870.0


## Step 6 : Convert to Matrix Format

### Objective

Convert the grouped monthly spending data into a matrix.


In [39]:
category_monthly_matrix = (
    category_monthly
    .pivot(
        index="Category",
        columns="Month_Year",
        values="Total_Spending"
    )
    .fillna(0)
)

In [40]:
month_order = [
    "Jan-2024",
    "Feb-2024",
    "Mar-2024",
    "Apr-2024",
    "May-2024",
    "Jun-2024"
]

category_monthly_matrix = category_monthly_matrix.reindex(
    columns=month_order
)

category_monthly_matrix

Month_Year,Jan-2024,Feb-2024,Mar-2024,Apr-2024,May-2024,Jun-2024
Category,,,,,,
Cafe,3690.0,4273.0,5448.0,6564.0,5668.0,5802.0
Cash Withdrawal,2000.0,5000.0,8000.0,5500.0,8000.0,17000.0
E-Commerce,98623.0,94011.0,108215.0,69219.0,95776.0,138033.0
Entertainment,4030.0,5102.0,5927.0,4392.0,1844.0,5469.0
Food Delivery,20890.0,21452.0,20850.0,23054.0,22167.0,20641.0
Fuel,30322.0,2079.0,26164.0,18718.0,9138.0,2882.0
Groceries,17649.0,7517.0,5523.0,6225.0,7862.0,5432.0
Housing,18000.0,18000.0,18000.0,18000.0,18000.0,18000.0
Investments,38432.0,15000.0,68644.0,54126.0,48628.0,23330.0


## Step 7 : Identify Highest Spending Category Each Month

### Objective

Determine which spending category recorded the highest expenditure in each month.



In [41]:
highest_category = category_monthly_matrix.idxmax()

highest_amount = category_monthly_matrix.max()

monthly_highest = pd.DataFrame({
    "Highest Category": highest_category,
    "Amount": highest_amount
})

monthly_highest

,Highest Category,Amount
Month_Year,,
Jan-2024,E-Commerce,98623.0
Feb-2024,E-Commerce,94011.0
Mar-2024,E-Commerce,108215.0
Apr-2024,E-Commerce,69219.0
May-2024,E-Commerce,95776.0
Jun-2024,E-Commerce,138033.0


## Step 8 : Monthly Trend Report

### Objective

Generate a text-based monthly spending report.



In [42]:
print("=" * 60)
print("MONTHLY SPENDING TREND REPORT")
print("=" * 60)

for month in month_order:
    if month in monthly_highest.index:
        category = monthly_highest.loc[month, "Highest Category"]
        amount = monthly_highest.loc[month, "Amount"]

        print(f"{month:<10} | {category:<20} | ₹{amount:,.2f}")

MONTHLY SPENDING TREND REPORT
Jan-2024   | E-Commerce           | ₹98,623.00
Feb-2024   | E-Commerce           | ₹94,011.00
Mar-2024   | E-Commerce           | ₹108,215.00
Apr-2024   | E-Commerce           | ₹69,219.00
May-2024   | E-Commerce           | ₹95,776.00
Jun-2024   | E-Commerce           | ₹138,033.00


## Step 9 : Analyze Monthly Spending Growth

### Objective

Calculate the month-to-month change in total monthly expenses.



In [43]:
monthly_growth = monthly_summary.copy()

monthly_growth["Monthly_Change"] = (
    monthly_growth["debit"].diff()
)

monthly_growth["Growth (%)"] = (
    monthly_growth["debit"]
    .pct_change()
    * 100
).round(2)

monthly_growth

Type,Date,Month_Year,credit,debit,Monthly_Change,Growth (%)
0,2024-01-01,Jan-2024,84728.0,290767.0,NaN,NaN
1,2024-02-01,Feb-2024,84724.0,234071.0,-56696.0,-19.50
2,2024-03-01,Mar-2024,84701.0,330458.0,96387.0,41.18
3,2024-04-01,Apr-2024,84736.0,251382.0,-79076.0,-23.93
4,2024-05-01,May-2024,85393.0,277306.0,25924.0,10.31
5,2024-06-01,Jun-2024,85492.0,294917.0,17611.0,6.35


In [44]:
print("=" * 70)
print("MONTHLY GROWTH REPORT")
print("=" * 70)

for _, row in monthly_growth.iterrows():

    if pd.isna(row["Monthly_Change"]):
        change = "N/A"
    elif row["Monthly_Change"] >= 0:
        change = f"+₹{row['Monthly_Change']:,.0f}"
    else:
        change = f"-₹{abs(row['Monthly_Change']):,.0f}"

    print(
        f"{row['Month_Year']:<10}"
        f" Expense: ₹{row['debit']:>9,.0f}"
        f"   Change: {change}"
    )

MONTHLY GROWTH REPORT
Jan-2024   Expense: ₹  290,767   Change: N/A
Feb-2024   Expense: ₹  234,071   Change: -₹56,696
Mar-2024   Expense: ₹  330,458   Change: +₹96,387
Apr-2024   Expense: ₹  251,382   Change: -₹79,076
May-2024   Expense: ₹  277,306   Change: +₹25,924
Jun-2024   Expense: ₹  294,917   Change: +₹17,611


### Observation

The Monthly Spending Trends feature successfully analyzed spending patterns across different months.

Key outcomes include:

- Generated month-wise income and expense summaries.
- Created a category-wise monthly spending matrix.
- Identified the highest spending category for each month.
- Calculated month-over-month changes in expenditure to highlight financial trends.
- Produced a text-based monthly trend report in accordance with the project constraints.

# Feature 6 : Time-of-Day Analysis

The Time-of-Day Analysis examines when financial transactions occur throughout the day.

By grouping transactions into different time periods, this feature helps identify spending habits based on the time of transaction.

## Step 1 : Extract Transaction Hour

### Objective

Extract the transaction hour from the timestamp.



In [45]:

# Feature 6 : Time-of-Day Analysis
# Step 1 : Extract Transaction Hour


df["Hour"] = pd.to_datetime(
    df["Time"],
    format="%H:%M"
).dt.hour

df[["Time", "Hour"]].head(10)

,Time,Hour
0,03:11,3
1,05:44,5
2,09:35,9
3,14:07,14
4,14:23,14
5,14:48,14
6,14:50,14
7,21:44,21
8,05:18,5
9,06:55,6


## Step 2 : Classify Time of Day

### Objective

Group every transaction into a time-of-day category.



In [46]:
def time_period(hour):

    if 5 <= hour < 12:
        return "Morning"

    elif 12 <= hour < 17:
        return "Afternoon"

    elif 17 <= hour < 21:
        return "Evening"

    else:
        return "Night"


df["Time_Period"] = df["Hour"].apply(time_period)

df[["Time", "Hour", "Time_Period"]].head(15)

,Time,Hour,Time_Period
0,03:11,3,Night
1,05:44,5,Morning
2,09:35,9,Morning
3,14:07,14,Afternoon
4,14:23,14,Afternoon
5,14:48,14,Afternoon
6,14:50,14,Afternoon
7,21:44,21,Night
8,05:18,5,Morning
9,06:55,6,Morning


## Step 3 : Analyze Transactions by Time Period

### Objective

Calculate the number of transactions and total spending for each time period.



In [47]:
time_summary = (
    df[df["Type"] == "debit"]
    .groupby("Time_Period")
    .agg(
        Transactions=("Amount", "count"),
        Total_Spending=("Amount", "sum")
    )
)

time_summary

,Transactions,Total_Spending
Time_Period,,
Afternoon,314,452169.0
Evening,340,363884.0
Morning,348,501130.0
Night,302,361718.0


## Step 4 : Sort Time Periods

### Objective

Arrange the time periods in chronological order.



In [48]:
time_order = [
    "Morning",
    "Afternoon",
    "Evening",
    "Night"
]

time_summary = time_summary.reindex(time_order)

time_summary

,Transactions,Total_Spending
Time_Period,,
Morning,348,501130.0
Afternoon,314,452169.0
Evening,340,363884.0
Night,302,361718.0


## Step 5 : Generate Time-of-Day Report

### Objective

Generate a formatted report summarizing spending during different times of the day.



In [49]:
print("=" * 65)
print("TIME OF DAY ANALYSIS REPORT")
print("=" * 65)

for period, row in time_summary.iterrows():

    print(
        f"{period:<10}"
        f" | Transactions: {int(row['Transactions']):>4}"
        f" | Spending: ₹{row['Total_Spending']:>10,.2f}"
    )

TIME OF DAY ANALYSIS REPORT
Morning    | Transactions:  348 | Spending: ₹501,130.00
Afternoon  | Transactions:  314 | Spending: ₹452,169.00
Evening    | Transactions:  340 | Spending: ₹363,884.00
Night      | Transactions:  302 | Spending: ₹361,718.00


### Observation

The Time-of-Day Analysis successfully categorized transactions into Morning, Afternoon, Evening, and Night based on the transaction hour.

Key outcomes include:

- Extracted transaction hours from the Time column.
- Classified each transaction into an appropriate time period.
- Calculated the number of transactions and total spending for each period.
- Generated a text-based summary report to highlight spending patterns throughout the day.

The analysis shows that the highest spending occurred during the **Morning** period.

# Feature 7 : Anomaly Detection

The Anomaly Detection feature identifies unusually large transactions within each spending category using the Z-score statistical method.

Instead of comparing every transaction globally, each transaction is compared only with other transactions belonging to the same category.

Transactions having a Z-score greater than 2 are considered anomalous.

## Step 1 : Calculate Category-wise Mean and Standard Deviation

### Objective

Calculate the average spending and standard deviation for every category.



In [50]:

# Feature 7
# Step 1 : Category Statistics

debit_df = df[df["Type"] == "debit"].copy()

debit_df["Category_Mean"] = (
    debit_df.groupby("Category")["Amount"]
    .transform("mean")
)

debit_df["Category_STD"] = (
    debit_df.groupby("Category")["Amount"]
    .transform("std")
)

debit_df.head()

,Date,Time,Description,Type,Amount,Balance,Mode,Ref,Vendor,Category,Month,Year,Month_Year,Hour,Time_Period,Category_Mean,Category_STD
0,2024-01-01,03:11,AMAZON SELLER SVCS,debit,2462.0,678275.0,UPI,TXN190872,Amazon,E-Commerce,Jan,2024,Jan-2024,3,Night,3510.912791,4522.129715
1,2024-01-01,05:44,BHIM-BMTC,debit,50.0,681007.0,UPI,TXN143064,BMTC,Transport,Jan,2024,Jan-2024,5,Morning,229.896000,147.398938
3,2024-01-01,14:07,UPI-AMAN-8934@OKAXIS,debit,1828.0,-748745.0,UPI,TXN569226,Personal Transfer,Personal Transfer,Jan,2024,Jan-2024,14,Afternoon,1417.071429,871.968283
4,2024-01-01,14:23,BHIM-BLINKIT,debit,270.0,680737.0,UPI,TXN968962,Blinkit,Quick Commerce,Jan,2024,Jan-2024,14,Afternoon,495.683938,199.088152
5,2024-01-01,14:48,BHIM ZEPTO,debit,625.0,677650.0,UPI,TXN370902,Zepto,Quick Commerce,Jan,2024,Jan-2024,14,Afternoon,495.683938,199.088152


## Step 2 : Calculate Z-Score

### Objective

Calculate the Z-score for every debit transaction.



In [51]:
debit_df["Z_Score"] = (
    (debit_df["Amount"] - debit_df["Category_Mean"])
    / debit_df["Category_STD"]
)

debit_df.head()

,Date,Time,Description,Type,Amount,Balance,Mode,Ref,Vendor,Category,Month,Year,Month_Year,Hour,Time_Period,Category_Mean,Category_STD,Z_Score
0,2024-01-01,03:11,AMAZON SELLER SVCS,debit,2462.0,678275.0,UPI,TXN190872,Amazon,E-Commerce,Jan,2024,Jan-2024,3,Night,3510.912791,4522.129715,-0.231951
1,2024-01-01,05:44,BHIM-BMTC,debit,50.0,681007.0,UPI,TXN143064,BMTC,Transport,Jan,2024,Jan-2024,5,Morning,229.896000,147.398938,-1.220470
3,2024-01-01,14:07,UPI-AMAN-8934@OKAXIS,debit,1828.0,-748745.0,UPI,TXN569226,Personal Transfer,Personal Transfer,Jan,2024,Jan-2024,14,Afternoon,1417.071429,871.968283,0.471266
4,2024-01-01,14:23,BHIM-BLINKIT,debit,270.0,680737.0,UPI,TXN968962,Blinkit,Quick Commerce,Jan,2024,Jan-2024,14,Afternoon,495.683938,199.088152,-1.133588
5,2024-01-01,14:48,BHIM ZEPTO,debit,625.0,677650.0,UPI,TXN370902,Zepto,Quick Commerce,Jan,2024,Jan-2024,14,Afternoon,495.683938,199.088152,0.649542


## Step 3 : Detect Anomalous Transactions

### Objective

Identify unusually large transactions.



In [52]:
anomalies = debit_df[
    debit_df["Z_Score"] > 2
].copy()

anomalies = anomalies.sort_values(
    by="Z_Score",
    ascending=False
)

print("Total Anomalies :", len(anomalies))

anomalies[
    [
        "Date",
        "Vendor",
        "Category",
        "Amount",
        "Z_Score"
    ]
].head(10)

Total Anomalies : 26


,Date,Vendor,Category,Amount,Z_Score
1298,2024-06-26,Amazon,E-Commerce,22008.0,4.090349
269,2024-02-07,Amazon,E-Commerce,21986.0,4.085484
414,2024-02-26,Restaurant,Restaurants,8383.0,3.884639
475,2024-03-05,Amazon,E-Commerce,19917.0,3.627956
1271,2024-06-22,Dineout,Restaurants,7935.0,3.627582
653,2024-03-31,Meghana Foods,Restaurants,7931.0,3.625287
464,2024-03-04,Truffles,Restaurants,7441.0,3.344131
49,2024-01-07,Restaurant,Restaurants,7314.0,3.271260
450,2024-03-02,Amazon,E-Commerce,18273.0,3.264410
1079,2024-05-27,Flipkart,E-Commerce,17831.0,3.166669


## Step 4 : Category-wise Anomaly Summary

### Objective

Summarize anomalies across spending categories.



In [53]:
category_anomalies = (
    anomalies.groupby("Category")
    .agg(
        Total_Anomalies=("Amount", "count"),
        Total_Anomaly_Amount=("Amount", "sum")
    )
    .sort_values(
        by="Total_Anomalies",
        ascending=False
    )
)

category_anomalies

,Total_Anomalies,Total_Anomaly_Amount
Category,,
E-Commerce,15,254211.0
Restaurants,5,39004.0
Entertainment,2,2908.0
Fuel,2,20913.0
Telecom,2,3332.0


## Step 5 : Anomaly Detection Report

### Objective

Generate a concise anomaly summary.



In [54]:
print("=" * 70)
print("ANOMALY DETECTION REPORT")
print("=" * 70)

print(f"Total Debit Transactions : {len(debit_df)}")
print(f"Total Anomalies          : {len(anomalies)}")

print("\nTop 5 Anomalies\n")

for _, row in anomalies.head(5).iterrows():
    print(
        f"{row['Date'].date()} | "
        f"{row['Vendor']:<20} | "
        f"{row['Category']:<18} | "
        f"₹{row['Amount']:>10,.2f} | "
        f"Z = {row['Z_Score']:.2f}"
    )

ANOMALY DETECTION REPORT
Total Debit Transactions : 1304
Total Anomalies          : 26

Top 5 Anomalies

2024-06-26 | Amazon               | E-Commerce         | ₹ 22,008.00 | Z = 4.09
2024-02-07 | Amazon               | E-Commerce         | ₹ 21,986.00 | Z = 4.09
2024-02-26 | Restaurant           | Restaurants        | ₹  8,383.00 | Z = 3.88
2024-03-05 | Amazon               | E-Commerce         | ₹ 19,917.00 | Z = 3.63
2024-06-22 | Dineout              | Restaurants        | ₹  7,935.00 | Z = 3.63


### Observation

The Z-score method successfully identified transactions that were significantly higher than the average spending within their respective categories.

Key outcomes include:

- Computed category-wise mean and standard deviation.
- Calculated the Z-score for every debit transaction.
- Flagged transactions with a Z-score greater than 2.
- Generated category-wise anomaly summaries and a formatted anomaly report.

This approach follows the project specification and provides category-specific anomaly detection.

In [55]:

# Savings Rate Calculation


total_income = df[df["Type"] == "credit"]["Amount"].sum()

total_expense = df[df["Type"] == "debit"]["Amount"].sum()

net_savings = total_income - total_expense

savings_rate = (net_savings / total_income) * 100

print(f"Total Income  : ₹{total_income:,.2f}")
print(f"Total Expense : ₹{total_expense:,.2f}")
print(f"Net Savings   : ₹{net_savings:,.2f}")
print(f"Savings Rate  : {savings_rate:.2f}%")

Total Income  : ₹509,774.00
Total Expense : ₹1,678,901.00
Net Savings   : ₹-1,169,127.00
Savings Rate  : -229.34%


# Feature 8 : Spending Archetype Detection

The Spending Archetype Detection feature analyzes Rahul's transaction history and assigns one or more financial personalities based on predefined spending rules.

Unlike traditional classification, a user can satisfy multiple archetypes simultaneously. Each detected archetype is supported with quantitative evidence from the spending analysis.

## Step 1 : Calculate Spending Metrics

### Objective

Calculate all metrics required for archetype detection.



In [56]:

# Feature 8 : Fresh Metrics


debit_df = df[df["Type"] == "debit"].copy()
credit_df = df[df["Type"] == "credit"].copy()

# Totals
total_income = credit_df["Amount"].sum()
total_expense = debit_df["Amount"].sum()
net_savings = total_income - total_expense

if total_income > 0:
    savings_rate = (net_savings / total_income) * 100
else:
    savings_rate = 0

# Category spending
category_spending = (
    debit_df.groupby("Category")["Amount"]
    .sum()
)

category_percentage = (
    category_spending / total_expense * 100
)

food_percentage = (
    category_percentage.get("Food Delivery", 0)
    + category_percentage.get("Restaurants", 0)
    + category_percentage.get("Cafe", 0)
)

quick_percentage = category_percentage.get("Quick Commerce", 0)

shopping_percentage = category_percentage.get("E-Commerce", 0)

investment_percentage = category_percentage.get("Investments", 0)

transport_percentage = category_percentage.get("Transport", 0)

print("Total Income :", total_income)
print("Total Expense :", total_expense)
print("Savings Rate :", round(savings_rate,2))

print("\nCategory Percentages\n")
print(category_percentage.round(2))

Total Income : 509774.0
Total Expense : 1678901.0
Savings Rate : -229.34

Category Percentages

Category
Cafe                  1.87
Cash Withdrawal       2.71
E-Commerce           35.97
Entertainment         1.59
Food Delivery         7.69
Fuel                  5.32
Groceries             2.99
Housing               6.43
Investments          14.78
Others                0.55
Personal Transfer     1.18
Quick Commerce        5.70
Restaurants           7.01
Telecom               0.93
Transport             3.42
Utilities             1.85
Name: Amount, dtype: float64


## Step 2 : Define Archetype Detection Functions

### Objective

Create independent functions for each spending archetype.



In [57]:

# Archetype Functions

def foodie():
    return food_percentage > 25


def quick_commerce_junkie():
    return quick_percentage > 15


def shopaholic():
    return shopping_percentage > 15


def investor():
    return investment_percentage > 15


def cab_commuter():
    return transport_percentage > 10


def yolo_spender():
    return savings_rate < 10


def disciplined_saver():
    return savings_rate > 40

## Step 3 : Late-Night Snacker Detection

### Objective

Identify whether the majority of Food Delivery transactions occur late at night.


In [58]:
food_orders = debit_df[
    debit_df["Category"] == "Food Delivery"
]

late_night_orders = food_orders[
    (food_orders["Hour"] >= 21)
    |
    (food_orders["Hour"] <= 2)
]

if len(food_orders) > 0:
    late_night_percentage = (
        len(late_night_orders)
        /
        len(food_orders)
    ) * 100
else:
    late_night_percentage = 0

## Step 4 : Subscription Lover Detection

### Objective

Determine whether Rahul actively maintains multiple digital subscriptions.



In [59]:
subscription_vendors = [
    "Netflix",
    "Spotify",
    "Disney+ Hotstar"
]

active_subscriptions = (
    debit_df[
        debit_df["Vendor"].isin(subscription_vendors)
    ]["Vendor"]
    .nunique()
)

In [60]:

# Late Night Snacker


food_orders = debit_df[
    debit_df["Category"] == "Food Delivery"
]

late_night_orders = food_orders[
    (food_orders["Hour"] >= 21)
    |
    (food_orders["Hour"] <= 2)
]

if len(food_orders) > 0:
    late_night_percentage = (
        len(late_night_orders) / len(food_orders)
    ) * 100
else:
    late_night_percentage = 0


def late_night_snacker():
    return late_night_percentage > 50

In [61]:
def foodie():
    return food_percentage > 25

def quick_commerce_junkie():
    return quick_percentage > 15

def shopaholic():
    return shopping_percentage > 15

def investor():
    return investment_percentage > 15

def cab_commuter():
    return transport_percentage > 10

def yolo_spender():
    return savings_rate < 10

def disciplined_saver():
    return savings_rate > 40

def subscription_lover():
    return active_subscriptions >= 5

## Step 5 : Detect All Matching Archetypes

### Objective

Evaluate every archetype rule and generate the final financial personality profile.


In [62]:
archetypes = []

if foodie():
    archetypes.append(
        f"THE FOODIE ({food_percentage:.2f}% on food)"
    )

if quick_commerce_junkie():
    archetypes.append(
        f"THE QUICK COMMERCE JUNKIE ({quick_percentage:.2f}% on Quick Commerce)"
    )

if shopaholic():
    archetypes.append(
        f"THE SHOPAHOLIC ({shopping_percentage:.2f}% on E-Commerce)"
    )

if investor():
    archetypes.append(
        f"THE INVESTOR ({investment_percentage:.2f}% on Investments)"
    )

if late_night_snacker():
    archetypes.append(
        f"THE LATE-NIGHT SNACKER ({late_night_percentage:.2f}% late-night food orders)"
    )

if subscription_lover():
    archetypes.append(
        f"THE SUBSCRIPTION LOVER ({active_subscriptions} active subscriptions)"
    )

if cab_commuter():
    archetypes.append(
        f"THE CAB COMMUTER ({transport_percentage:.2f}% transport spending)"
    )

if yolo_spender():
   archetypes.append(
    f"THE YOLO SPENDER (Expenses exceed income by {abs(savings_rate):.2f}%)"
)

if disciplined_saver():
    archetypes.append(
        f"THE DISCIPLINED SAVER (Savings Rate = {savings_rate:.2f}%)"
    )

In [63]:
print("=" * 75)
print("RAHUL'S SPENDING ARCHETYPE ANALYSIS")
print("=" * 75)

checks = [
    ("THE FOODIE", foodie(), f"{food_percentage:.2f}%"),
    ("THE QUICK COMMERCE JUNKIE", quick_commerce_junkie(), f"{quick_percentage:.2f}%"),
    ("THE SHOPAHOLIC", shopaholic(), f"{shopping_percentage:.2f}%"),
    ("THE INVESTOR", investor(), f"{investment_percentage:.2f}%"),
    ("THE LATE-NIGHT SNACKER", late_night_snacker(), f"{late_night_percentage:.2f}%"),
    ("THE SUBSCRIPTION LOVER", subscription_lover(), str(active_subscriptions)),
    ("THE CAB COMMUTER", cab_commuter(), f"{transport_percentage:.2f}%"),
    ("THE YOLO SPENDER", yolo_spender(), f"{savings_rate:.2f}%"),
    ("THE DISCIPLINED SAVER", disciplined_saver(), f"{savings_rate:.2f}%")
]

for name, result, metric in checks:
    status = "MATCHED" if result else "NOT MATCHED"
    print(f"{name:<32} {status:<12} ({metric})")

print("\n" + "=" * 75)
print("FINAL DETECTED ARCHETYPES")
print("=" * 75)

if archetypes:
    for archetype in archetypes:
        print(f"✓ {archetype}")
else:
    print("No archetypes detected.")

RAHUL'S SPENDING ARCHETYPE ANALYSIS
THE FOODIE                       NOT MATCHED  (16.57%)
THE QUICK COMMERCE JUNKIE        NOT MATCHED  (5.70%)
THE SHOPAHOLIC                   MATCHED      (35.97%)
THE INVESTOR                     NOT MATCHED  (14.78%)
THE LATE-NIGHT SNACKER           NOT MATCHED  (21.21%)
THE SUBSCRIPTION LOVER           NOT MATCHED  (3)
THE CAB COMMUTER                 NOT MATCHED  (3.42%)
THE YOLO SPENDER                 MATCHED      (-229.34%)
THE DISCIPLINED SAVER            NOT MATCHED  (-229.34%)

FINAL DETECTED ARCHETYPES
✓ THE SHOPAHOLIC (35.97% on E-Commerce)
✓ THE YOLO SPENDER (Expenses exceed income by 229.34%)


# Final SpendDNA Report

The Final SpendDNA Report combines the results of all previous analyses into a single comprehensive summary.

This report highlights the user's spending behaviour, financial trends, anomalies, and spending archetypes, providing an overall financial profile.

In [64]:
# ==========================================================
#                 FINAL SPENDDNA REPORT
# ==========================================================

print("=" * 90)
print(" " * 30 + "SPENDDNA ANALYSIS REPORT")
print("=" * 90)

# ==========================================================
# Executive Summary
# ==========================================================

print("\n1. EXECUTIVE SUMMARY")
print("-" * 90)

print(f"Total Transactions : {len(df)}")
print(f"Total Income       : ₹{total_income:,.2f}")
print(f"Total Expenses     : ₹{total_expense:,.2f}")
print(f"Net Savings        : ₹{net_savings:,.2f}")
print(f"Savings Rate       : {savings_rate:.2f}%")

# ==========================================================
# Top Spending Categories
# ==========================================================

print("\n2. TOP SPENDING CATEGORIES")
print("-" * 90)

top_categories = (
    category_spending
    .sort_values(ascending=False)
    .head(5)
)

for category, amount in top_categories.items():

    percentage = (amount / total_expense) * 100

    print(
        f"{category:<25}"
        f"₹ {amount:>12,.2f}"
        f"   ({percentage:.2f}%)"
    )

# ==========================================================
# Monthly Spending
# ==========================================================

print("\n3. MONTHLY SPENDING TREND")
print("-" * 90)

monthly_report = (
    debit_df
    .groupby(["Year", "Month", "Month_Year"], as_index=False)["Amount"]
    .sum()
    .sort_values(by=["Year", "Month"])
)

for _, row in monthly_report.iterrows():

    print(
        f"{row['Month_Year']:<15}"
        f"₹ {row['Amount']:>12,.2f}"
    )

highest_month = monthly_report.loc[
    monthly_report["Amount"].idxmax()
]

print(
    f"\nHighest Spending Month : "
    f"{highest_month['Month_Year']} "
    f"(₹{highest_month['Amount']:,.2f})"
)

# ==========================================================
# Time of Day Analysis
# ==========================================================

print("\n4. TIME OF DAY ANALYSIS")
print("-" * 90)

time_report = (
    debit_df
    .groupby("Time_Period")
    .agg(
        Transactions=("Amount", "count"),
        Spending=("Amount", "sum")
    )
)

for period, row in time_report.iterrows():

    print(
        f"{period:<12}"
        f"{int(row['Transactions']):>5} Transactions"
        f"    ₹ {row['Spending']:>12,.2f}"
    )

# ==========================================================
# Anomaly Summary
# ==========================================================

print("\n5. ANOMALY SUMMARY")
print("-" * 90)

print(f"Total Anomalies Detected : {len(anomalies)}")

print("\nTop 5 Anomalous Transactions\n")

for _, row in anomalies.head(5).iterrows():

    print(
        f"{row['Date'].date()} | "
        f"{row['Vendor']:<18} | "
        f"{row['Category']:<18} | "
        f"₹ {row['Amount']:>8,.2f} | "
        f"Z = {row['Z_Score']:.2f}"
    )

# ==========================================================
# Spending Archetypes
# ==========================================================

print("\n6. DETECTED SPENDING ARCHETYPES")
print("-" * 90)

if len(archetypes) == 0:

    print("No Spending Archetypes Detected.")

else:

    for archetype in archetypes:

        print(f"✓ {archetype}")

# ==========================================================
# Key Insights
# ==========================================================

print("\n7. KEY INSIGHTS")
print("-" * 90)

highest_category = category_spending.idxmax()

highest_category_amount = category_spending.max()

highest_category_percentage = (
    highest_category_amount / total_expense
) * 100

highest_time = (
    time_report["Spending"]
    .idxmax()
)

print(
    f"• {highest_category} is the highest spending category "
    f"with ₹{highest_category_amount:,.2f} "
    f"({highest_category_percentage:.2f}% of total spending)."
)

print(
    f"• {len(anomalies)} anomalous transactions were detected. "
    f"Amazon contributes the highest-value anomalies."
)

print(
    f"• Most spending occurs during the "
    f"{highest_time.lower()} period."
)

# ==========================================================
# Recommendations
# ==========================================================

print("\n8. RECOMMENDATIONS")
print("-" * 90)

print(
    "• Review high-value E-Commerce purchases regularly."
)

print(
    "• Maintain investments while reducing discretionary spending."
)

print(
    "• Monitor anomalous transactions to identify unusual expenses."
)

print(
    "• Aim to improve the savings rate through better budgeting."
)

print("\n" + "=" * 90)
print(" " * 35 + "END OF REPORT")
print("=" * 90)

                              SPENDDNA ANALYSIS REPORT

1. EXECUTIVE SUMMARY
------------------------------------------------------------------------------------------
Total Transactions : 1310
Total Income       : ₹509,774.00
Total Expenses     : ₹1,678,901.00
Net Savings        : ₹-1,169,127.00
Savings Rate       : -229.34%

2. TOP SPENDING CATEGORIES
------------------------------------------------------------------------------------------
E-Commerce               ₹   603,877.00   (35.97%)
Investments              ₹   248,160.00   (14.78%)
Food Delivery            ₹   129,054.00   (7.69%)
Restaurants              ₹   117,737.00   (7.01%)
Housing                  ₹   108,000.00   (6.43%)

3. MONTHLY SPENDING TREND
------------------------------------------------------------------------------------------
Apr-2024       ₹   251,382.00
Feb-2024       ₹   234,071.00
Jan-2024       ₹   290,767.00
Jun-2024       ₹   294,917.00
Mar-2024       ₹   330,458.00
May-2024       ₹   277,306.00

Hi

## Observation

The Final SpendDNA Report successfully consolidated the results of all analytical features into a unified financial summary.

The report includes:

- Executive financial summary
- Category-wise spending overview
- Monthly spending trends
- Time-of-day spending analysis
- Anomaly detection summary
- Spending archetype analysis
- Key insights
- Financial recommendations

This report provides a concise and comprehensive overview of the user's spending behaviour and financial patterns.

# Reflection

The SpendDNA project provided hands-on experience in analyzing financial transaction data using Python and Pandas. Throughout the project, I learned how to clean raw banking data, extract vendors, categorize transactions, analyze monthly and time-based spending patterns, detect anomalies using Z-score analysis, and identify spending archetypes using rule-based logic.

The project strengthened my understanding of data preprocessing, statistical analysis, aggregation techniques, and report generation. It also demonstrated how raw financial transaction data can be transformed into meaningful insights that support better financial decision-making.

Overall, this project improved my practical data analytics skills and enhanced my ability to build end-to-end analytical solutions using Python.